### PR TITLE
HAAR-588:  token endpoint

### PR DESCRIPTION
**Pull request by ravmoj**
RESET token endpoint in auth

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/TokenController.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.web.bind.annotation.RestController
+import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken
+import uk.gov.justice.digital.hmpps.oauth2server.model.CreateTokenRequest
+import uk.gov.justice.digital.hmpps.oauth2server.verify.TokenService
+
+@RestController
+@Validated
+class TokenController(private val tokenService: TokenService) {
+
+  @PreAuthorize("hasRole('ROLE_CREATE_USER')")
```
Comment (Mjwillis) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/TokenController.kt:16]: This should be ROLE_CREATE_EMAIL_TOKEN

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/verify/TokenService.kt [hunk 2]
```diff
@@ -79,6 +81,20 @@ class TokenService(
...
+  fun createTokenForNewUser(tokenType: TokenType, username: String, email: String, source: AuthSource): String {
+    log.info("Requesting {} for {}", tokenType.description, username)
+    val user = userService.createUser(username, email, source).orElseThrow()
+    val userToken = user.createToken(tokenType)
+    telemetryClient.trackEvent(
+      "${tokenType.description}Request",
+      mapOf("username" to username),
+      null
+    )
+    userToken.tokenExpiry = LocalDateTime.now().plusDays(7)
```
Comment (Mjwillis) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/verify/TokenService.kt:94]: maybe make configurable as a param in request with default of 7d

File: src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/TokenControllerIntTest.kt [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+  @Autowired
+  private lateinit var userTokenRepository: UserTokenRepository
+
+  @Test
+  fun `CreateToken based on user details provided`() {
+
+    val tokenFromRestEndPoint = webTestClient
+      .post().uri("/api/new-token")
+      .headers(setAuthorisation("AUTH_ADM", listOf("ROLE_CREATE_USER")))
```
Comment (Mjwillis) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/TokenControllerIntTest.kt:27]: See role change

**Review by Mjwillis — CHANGES_REQUESTED** [review]
Change the role name

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/TokenController.kt (hunk 1 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/TokenController.kt]: Updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/verify/TokenService.kt (hunk 2 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/verify/TokenService.kt]: Updated

**Review by Mjwillis — APPROVED** [review]
LGTM

**Pull request merged**